File: packages/dotnet/src/Models/SendEmail.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Axene.Mailer
{
    /// <summary>
    /// A message to send. <see cref="From"/>, <see cref="To"/>, and
    /// <see cref="Subject"/> are required; provide <see cref="Html"/>,
    /// <see cref="Text"/>, or both.
    /// </summary>
    public sealed class SendEmail
    {
        /// <summary>Sender address. Must be on a verified domain in your account.</summary>
        public Address From { get; set; } = new Address();

        /// <summary>One or more recipients.</summary>
        public List<Address> To { get; set; } = new List<Address>();

        /// <summary>Subject line.</summary>
        public string Subject { get; set; } = "";

        /// <summary>HTML body.</summary>
        public string? Html { get; set; }

        /// <summary>Plain-text body.</summary>
        public string? Text { get; set; }

        /// <summary>Carbon-copy recipients.</summary>
        public List<Address>? Cc { get; set; }

        /// <summary>Blind carbon-copy recipients.</summary>
        public List<Address>? Bcc { get; set; }

        /// <summary>Reply-to address.</summary>
        public Address? ReplyTo { get; set; }

        /// <summary>Custom headers to attach to the message.</summary>
        public Dictionary<string, string>? Headers { get; set; }

        /// <summary>Tags for filtering and analytics.</summary>
        public List<string>? Tags { get; set; }

        /// <summary>Schedule delivery for later (Starter plan and up).</summary>
        public DateTimeOffset? SendAt { get; set; }

        /// <summary>File attachments.</summary>
        public List<Attachment>? Attachments { get; set; }

        /// <summary>
        /// Build the JSON wire body. The API names the sender field <c>from_</c>,
        /// so the mapping from <see cref="From"/> happens here in one place, and
        /// null fields are omitted.
        /// </summary>
        internal Dictionary<string, object?> ToWire()
        {
            var d = new Dictionary<string, object?>
            {
                ["from_"] = From,
                ["to"] = To,
                ["subject"] = Subject,
                ["html"] = Html,
                ["text"] = Text,
                ["cc"] = Cc,
                ["bcc"] = Bcc,
                ["reply_to"] = ReplyTo,
                ["headers"] = Headers,
                ["tags"] = Tags,
                ["send_at"] = SendAt?.ToUniversalTime().ToString("o"),
                ["attachments"] = Attachments,
            };
            return d.Where(kv => kv.Value != null).ToDictionary(kv => kv.Key, kv => kv.Value);
        }
    }
}

File: packages/java/src/main/java/io/axene/mailer/SendEmail.java
package io.axene.mailer;

import java.time.Instant;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/** A message to send. Build with {@link #builder()}. {@code from}, {@code to}, and {@code subject} are required. */
public final class SendEmail {

    private final Address from;
    private final List<Address> to;
    private final String subject;
    private final String html;
    private final String text;
    private final List<Address> cc;
    private final List<Address> bcc;
    private final Address replyTo;
    private final Map<String, String> headers;
    private final List<String> tags;
    private final Instant sendAt;
    private final List<Attachment> attachments;

    private SendEmail(Builder b) {
        this.from = b.from;
        this.to = b.to;
        this.subject = b.subject;
        this.html = b.html;
        this.text = b.text;
        this.cc = b.cc;
        this.bcc = b.bcc;
        this.replyTo = b.replyTo;
        this.headers = b.headers;
        this.tags = b.tags;
        this.sendAt = b.sendAt;
        this.attachments = b.attachments;
    }

    public static Builder builder() {
        return new Builder();
    }

    /** Build the JSON wire form. The sender field is wire-named {@code from_}. */
    Map<String, Object> toWire() {
        if (from == null) throw new IllegalStateException("from is required");
        if (to == null || to.isEmpty()) throw new IllegalStateException("at least one `to` recipient is required");
        if (subject == null) throw new IllegalStateException("subject is required");
        Map<String, Object> m = new LinkedHashMap<>();
        m.put("from_", from.toMap());
        m.put("to", mapAddresses(to));
        m.put("subject", subject);
        if (html != null) m.put("html", html);
        if (text != null) m.put("text", text);
        if (cc != null) m.put("cc", mapAddresses(cc));
        if (bcc != null) m.put("bcc", mapAddresses(bcc));
        if (replyTo != null) m.put("reply_to", replyTo.toMap());
        if (headers != null) m.put("headers", headers);
        if (tags != null) m.put("tags", tags);
        if (sendAt != null) m.put("send_at", sendAt.toString());
        if (attachments != null) {
            List<Map<String, Object>> a = new ArrayList<>();
            for (Attachment at : attachments) a.add(at.toMap());
            m.put("attachments", a);
        }
        return m;
    }

    private static List<Map<String, Object>> mapAddresses(List<Address> list) {
        List<Map<String, Object>> out = new ArrayList<>();
        for (Address a : list) out.add(a.toMap());
        return out;
    }

    /** Fluent builder for {@link SendEmail}. */
    public static final class Builder {
        private Address from;
        private final List<Address> to = new ArrayList<>();
        private String subject;
        private String html;
        private String text;
        private List<Address> cc;
        private List<Address> bcc;
        private Address replyTo;
        private Map<String, String> headers;
        private List<String> tags;
        private Instant sendAt;
        private List<Attachment> attachments;

        public Builder from(String email) { this.from = new Address(email); return this; }
        public Builder from(String email, String name) { this.from = new Address(email, name); return this; }
        public Builder from(Address address) { this.from = address; return this; }

        public Builder to(String email) { this.to.add(new Address(email)); return this; }
        public Builder to(String email, String name) { this.to.add(new Address(email, name)); return this; }
        public Builder to(Address address) { this.to.add(address); return this; }

        public Builder cc(String email) { (cc = cc == null ? new ArrayList<>() : cc).add(new Address(email)); return this; }
        public Builder bcc(String email) { (bcc = bcc == null ? new ArrayList<>() : bcc).add(new Address(email)); return this; }

        public Builder subject(String subject) { this.subject = subject; return this; }
        public Builder html(String html) { this.html = html; return this; }
        public Builder text(String text) { this.text = text; return this; }
        public Builder replyTo(String email) { this.replyTo = new Address(email); return this; }
        public Builder header(String name, String value) {
            (headers = headers == null ? new LinkedHashMap<>() : headers).put(name, value);
            return this;
        }
        public Builder tag(String tag) { (tags = tags == null ? new ArrayList<>() : tags).add(tag); return this; }
        /** Schedule for later (Starter plan and up). */
        public Builder sendAt(Instant when) { this.sendAt = when; return this; }
        public Builder attachment(Attachment a) {
            (attachments = attachments == null ? new ArrayList<>() : attachments).add(a);
            return this;
        }

        public SendEmail build() { return new SendEmail(this); }
    }
}
